packages/circuit-ui/components/Modal/Modal.tsx

```diff
@@ -13,192 +13,217 @@
+import { ModalComponent, BaseModalProps } from '../ModalContext/ModalContext';
+import CloseButton from '../CloseButton';
+
+const TRANSITION_DURATION_MOBILE = 120;
+const TRANSITION_DURATION_DESKTOP = 240;
+const TRANSITION_DURATION = Math.max(
+  TRANSITION_DURATION_MOBILE,
+  TRANSITION_DURATION_DESKTOP,
+);
+
+const closeButtonStyles = (theme: Theme) => css`
+  position: absolute;
+  top: ${theme.spacings.byte};
+  right: ${theme.spacings.byte};
+
+  ${theme.mq.kilo} {
+    top: ${theme.spacings.mega};
+    right: ${theme.spacings.mega};
+  }
+`;
+
+export interface ModalProps extends BaseModalProps {
   /**
-   * Determines if the modal is visible or not.
+   * TODO: Add description
    */
-  isOpen?: boolean;
+  children:
+    | ReactNode
+    | (({ onClose }: { onClose: BaseModalProps['onClose'] }) => ReactNode);
   /**
-   * Function to close the modal. Passed down to the children
-   * render prop.
+   * TODO: Add description
    */
-  onClose?: OnClose;
+  variant: 'immersive' | 'contextual';
   /**
-   * React Modal's accessibility string.
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
    */
-  contentLabel?: string;
+  labelCloseButton: string;
   /**
-   * The element that should be used as root for the
-   * React portal used to display the modal. See
-   * http://reactcommunity.org/react-modal/accessibility/#app-element
+   * TODO: Add description. Default true.
    */
-  appElement?: string | HTMLElement;
-  /**
-   * Additional data that is dispatched with the tracking event.
-   */
-  tracking?: TrackingProps;
+  dismissible?: boolean;
+  className?: string;
 }
 
-export const TRANSITION_DURATION = 200;
-export const DEFAULT_APP_ELEMENT = '#root';
-
-const TOP_MARGIN = '10vh';
-const TRANSFORM_Y_FLOATING = '10vh';
-const FLOATING_TRANSITION = `${TRANSITION_DURATION}ms ease-in-out`;
-// eslint-disable-next-line max-len
-const FIXED_TRANSITION = `${TRANSITION_DURATION}ms cubic-bezier(0, 0.37, 0.64, 1)`;
-
 /**
- * Circuit UI's wrapper component for ReactModal. Uses the Card component
- * to wrap content passed as the children prop. Don't forget to set
- * the aria prop when using this.
+ * Circuit UI's wrapper component for ReactModal.
  * http://reactcommunity.org/react-modal/accessibility/#aria
  */
-export const Modal: FC<ModalProps> = ({
+export const Modal: ModalComponent<ModalProps> = ({
   children,
   onClose,
-  contentLabel = 'Modal',
-  appElement = DEFAULT_APP_ELEMENT,
-  isOpen = true,
+  variant,
+  dismissible = true,
+  labelCloseButton,
   tracking = {},
+  className,
   ...props
 }) => {
-  const theme: Theme = useTheme();
-  const handleClose =
-    useClickHandler(onClose, tracking, 'modal-close') || onClose;
-  ReactModal.setAppElement(appElement);
+  const handleClose = useClickHandler(onClose, tracking, 'modal-close');
   return (
-    <ClassNames>
-      {({ css }) => {
+    <ClassNames<Theme>>
+      {({ css: cssString, cx, theme }) => {
```

Comment:
The "normal" `css` is also imported [here](https://github.com/sumup-oss/circuit-ui/pull/949/files/cbd432c654a9a6a29ed447c77ee2ad7a418b1566#diff-c88f3430c369a6ab32984de0f262db27d0e39b955c4034a8ec1ca1d16180a4c6R17) so I had to rename it to avoid a conflict. I chose `cssString` because this function returns a classname (as opposed to the style object that's returned by the "normal" `css` function).